### PR TITLE
New/skill endpoints task #81 and #42

### PR DESCRIPTION
**Pull request by twistedomelette**
Made the interaction of the front and the back, you can delete, add, change the skill, as well as sort by the criterion

### CODE REVIEW

File: frontend/src/components/skills/rating/skill-rating.tsx [hunk 1]
```diff
@@ -45,12 +45,17 @@ const SkillElement = (props: SkillTypes): React.ReactElement => {
...
-          id: id,
-          name: nameSkill,
-          userId: user.id,
-          rating: ratingValues,
-        }),
+        skillActions.updateSkill([
+          {
+            id: id,
+            name: nameSkill,
+            type: 'Soft skills',
```
Comment (aleksandr-kuharenko) [inline frontend/src/components/skills/rating/skill-rating.tsx:52]: remove hardcode

File: backend/src/services/skill.service.ts [hunk 2]
```diff
@@ -1,10 +1,203 @@
...
+    });
+
+  await userSkillInstance.remove();
+
+  return { success: true, message: 'Skill deleted successfully' };
+};
+
+export const updateSkill = async (
+  id: Skill['id'],
+  body: any,
```
Comment (aleksandr-kuharenko) [inline backend/src/services/skill.service.ts]: do not use any

File: backend/src/services/skill.service.ts [hunk 3]
```diff
@@ -1,10 +1,203 @@
...
+    id: companyId,
+  });
+  const user = await userRepository.findOne({
+    where: { id: userId },
+  });
+  const allSkills: Skill[] = await skillRepository.find({
+    company: companyInstance,
+  } as FindManyOptions);
+
+  const skills: any = [];
```
Comment (aleksandr-kuharenko) [inline backend/src/services/skill.service.ts]: any

File: backend/src/services/skill.service.ts [hunk 4]
```diff
@@ -1,10 +1,203 @@
...
+  });
+  const user = await userRepository.findOne({
+    where: { id: userId },
+  });
+
+  const skills: any = [];
+  await asyncForEach(async ({ name, type }: Skill) => {
+    const skill = skillRepository.create({
+      name: name,
+      type: type,
```
Comment (aleksandr-kuharenko) [inline backend/src/services/skill.service.ts]:
```suggestion
      name,
      type,
```

File: backend/src/services/skill.service.ts [hunk 5]
```diff
@@ -1,10 +1,203 @@
...
+  });
+
+  const newSkills: Skill[] = userSkillInstance.map((userSkill) =>
+    skillsMapper(userSkill.skill),
+  );
+  const skillsWithRating = newSkills.map((el: Skill, index: number) => {
+    const selfRating = userSkillInstance[index].selfRating;
+    const mentorRating = userSkillInstance[index].mentorRating;
+    const reviewRating = userSkillInstance[index].reviewRating;
+    const isStarred = userSkillInstance[index].isStarred;
```
Comment (aleksandr-kuharenko) [inline backend/src/services/skill.service.ts]:
```suggestion
    const { selfRating, mentorRating, reviewRating, isStarred  } = userSkillInstance[index];
```

File: backend/src/services/skill.service.ts (hunk 2 above)
Comment (aleksandr-kuharenko) [reply backend/src/services/skill.service.ts]: use word 'data' instead of 'body'

File: backend/src/services/skill.service.ts [hunk 6]
```diff
@@ -1,10 +1,203 @@
...
+import { HttpCode, HttpError } from 'growup-shared';
+import UserRepository from '~/data/repositories/user.repository';
+import UserSkillRepository from '~/data/repositories/userskill.repository';
+
+export const getSkills = async (id: Company['id']): Promise<Skill[]> => {
+  const skillRepository = getCustomRepository(SkillRepository);
+  const companyRepository = getCustomRepository(CompanyRepository);
+
+  const companyInstance: Company = await companyRepository.findOne({
+    id: id,
```
Comment (aleksandr-kuharenko) [inline backend/src/services/skill.service.ts]:
```suggestion
    id,
```

File: backend/src/api/routes/company-route.ts [hunk 7]
```diff
@@ -31,7 +31,7 @@ const router: Router = Router();
 
 router.post(
   '/',
-  validatePermissions([RoleType.ADMIN]),
+  // validatePermissions([RoleType.ADMIN]),
```
Comment (aleksandr-kuharenko) [inline backend/src/api/routes/company-route.ts]:
```suggestion
  validatePermissions([RoleType.ADMIN]),
```

File: backend/src/services/skill.service.ts [hunk 8]
```diff
@@ -1,10 +1,203 @@
...
+    const skillRepository = getCustomRepository(SkillRepository);
+    const skillInstance = await skillRepository.findOne(id);
+    const userSkillRepository = getCustomRepository(UserSkillRepository);
+    const userSkillInstance = await userSkillRepository.findOne({
+      where: {
+        skill: id,
+      },
+      relations: ['user', 'skill'],
+    });
+    const ratings = [body[1].rating[0], body[1].rating[1], body[1].rating[2]];
```
Comment (aleksandr-kuharenko) [inline backend/src/services/skill.service.ts]: unreadable

**Review by aleksandr-kuharenko — CHANGES_REQUESTED** [review]

File: backend/src/services/skill.service.ts [hunk 9]
```diff
@@ -1,10 +1,218 @@
...
+import { SuccessResponse } from '~/common/models/responses/success';
+import { HttpCode, HttpError } from 'growup-shared';
+import UserRepository from '~/data/repositories/user.repository';
+import UserSkillRepository from '~/data/repositories/userskill.repository';
+import { SkillType } from '~/common/enums/skill-type';
+import { UserSkill } from '~/data/entities/user-skill';
+
+interface ISkill {
+  rating: (string | number)[];
+  isStarred: boolean | any;
```
Comment (nikita-remeslov) [inline backend/src/services/skill.service.ts]: any? We don't use any

File: backend/src/services/skill.service.ts [hunk 10]
```diff
@@ -1,10 +1,218 @@
...
+      ...el,
+      rating: [
+        selfRating ? selfRating : '',
+        mentorRating ? mentorRating : '',
+        reviewRating ? reviewRating : '',
+      ],
+      isStarred: isStarred,
+    };
+  });
+  console.log(skillsWithRating);
```
Comment (nikita-remeslov) [inline backend/src/services/skill.service.ts]: remove console.log

File: backend/src/services/skill.service.ts [hunk 11]
```diff
@@ -1,10 +1,218 @@
...
+  const userSkillRepository = getCustomRepository(UserSkillRepository);
+
+  const companyInstance: Company = await companyRepository.findOne({
+    id: companyId,
+  });
+  const user = await userRepository.findOne({
+    where: { id: userId },
+  });
+
+  const skills: any = [];
```
Comment (nikita-remeslov) [inline backend/src/services/skill.service.ts]: any

File: backend/src/services/skill.service.ts [hunk 12]
```diff
@@ -1,10 +1,218 @@
...
+export const connectSkills = async (
+  data: Skill[],
+  userId: string,
+  companyId: string,
+): Promise<SkillsCreationResponse> => {
+  const skillRepository = getCustomRepository(SkillRepository);
+  const companyRepository = getCustomRepository(CompanyRepository);
+  const userRepository = getCustomRepository(UserRepository);
+  const userSkillRepository = getCustomRepository(UserSkillRepository);
+  console.log('s');
```
Comment (nikita-remeslov) [inline backend/src/services/skill.service.ts]: console.log

File: backend/src/services/skill.service.ts [hunk 13]
```diff
@@ -1,10 +1,218 @@
...
+    const importantSkill = allSkills.find((skill) => skill.name === name);
+    const userSkill = userSkillRepository.create({
+      skill: importantSkill,
+      user,
+    });
+    await userSkill.save();
+    skills.push(importantSkill);
+  }, data);
+  await user.save();
+  console.log(skills);
```
Comment (nikita-remeslov) [inline backend/src/services/skill.service.ts]: console.log

File: backend/src/services/skill.service.ts [hunk 14]
```diff
@@ -1,10 +1,218 @@
...
+    });
+
+  await userSkillInstance.remove();
+
+  return { success: true, message: 'Skill deleted successfully' };
+};
+
+export const updateSkill = async (
+  id: Skill['id'],
+  data: any,
```
Comment (nikita-remeslov) [inline backend/src/services/skill.service.ts]: any

File: backend/src/services/skill.service.ts [hunk 15]
```diff
@@ -1,10 +1,218 @@
...
+    const skillRepository = getCustomRepository(SkillRepository);
+    const skillInstance = await skillRepository.findOne(id);
+    const userSkillRepository = getCustomRepository(UserSkillRepository);
+    const userSkillInstance = await userSkillRepository.findOne({
+      where: {
+        skill: id,
+      },
+      relations: ['user', 'skill'],
+    });
+    console.log(data);
```
Comment (nikita-remeslov) [inline backend/src/services/skill.service.ts]: console.log

File: frontend/src/common/interfaces/skill/skill.ts [hunk 16]
```diff
@@ -0,0 +1,12 @@
+export interface ISkill {
+  id: string;
+  name: string;
+  type?: string;
+  company?: string | undefined;
+  rating?: any;
```
Comment (nikita-remeslov) [inline frontend/src/common/interfaces/skill/skill.ts]: any

File: frontend/src/store/company/actions.ts [hunk 17]
```diff
@@ -9,7 +9,7 @@ const get_allCompanisesAsync = createAsyncThunk(
   ActionType.GET_ALL_COMPANIES,
   async (_, { dispatch }) => {
     const result = await companyApi.getAllCompamies();
-
+    console.warn(result);
```
Comment (nikita-remeslov) [inline frontend/src/store/company/actions.ts]: Do you need it?

File: frontend/src/store/skill/slice.ts [hunk 18]
```diff
@@ -1,45 +1,18 @@
 import { createSlice, PayloadAction } from '@reduxjs/toolkit';
 import { ReducerName } from 'common/enums/app/reducer-name.enum';
 import { ISkill } from 'components/skills/common/interfaces';
+// import { ISkill } from 'common/interfaces/skill/skill';
```
Comment (nikita-remeslov) [inline frontend/src/store/skill/slice.ts]: remove if you don't need this

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

**Review by aleksandr-kuharenko — APPROVED** [review]

**Review by AndriyCh1 — APPROVED** [review]

**Review by nikita-remeslov — APPROVED** [review]

**Pull request merged**